client/src/controllers/EnableDirtyFormCheckController.ts

```diff
@@ -0,0 +1,190 @@
+        this.commentApp.store.getState(),
+      );
+      this.commentApp.store.subscribe(() => {
+        // Update on a timeout to match the timings for responding to page form changes
+        clearTimeout(this.isDirtyTimeout);
+        this.isCommentsDirtyTimeout = setTimeout(
+          () => {
+            const newIsCommentsDirty = this.commentApp.selectors.selectIsDirty(
+              this.commentApp.store.getState(),
+            );
+            if (newIsCommentsDirty !== this.isCommentsDirty) {
+              this.isCommentsDirty = newIsCommentsDirty;
+              this.updateCallback(this.isDirty, this.isCommentsDirty);
+            }
+          },
+          this.isCommentsDirty ? 3000 : 300,
+        );
+      });
+    }
+
+    this.updateCallback(this.isDirty, this.isCommentsDirty);
+    this.isDirtyTimeout = -1;
+
+    if (!this.alwaysDirtyValue) {
+      setTimeout(() => {
+        this.initialData = this.getFormData();
+
+        const updateDirtyCheck = () => {
+          clearTimeout(this.isDirtyTimeout);
+          // If the form is dirty, it is relatively unlikely to become clean again, so
+          // run the dirty check on a relatively long timer that we reset on any form update
+          // otherwise, use a short timer both for nicer UX and to ensure widgets
+          // like Draftail have time to serialize their data
+          this.isDirtyTimeout = setTimeout(
+            this.updateIsDirty.bind(this),
+            this.isDirty ? 3000 : 300,
+          );
+        };
+
+        this.element.addEventListener('change', updateDirtyCheck);
+        this.element.addEventListener('keyup', updateDirtyCheck);
+        updateDirtyCheck();
+
+        const isValidInputNode = (node) =>
+          node.nodeType === node.ELEMENT_NODE &&
+          ['INPUT', 'TEXTAREA', 'SELECT'].includes(node.tagName);
+
+        this.observer = new MutationObserver((mutationList) => {
+          const hasMutationWithValidInputNode = mutationList.some(
+            (mutation) =>
+              Array.from(mutation.addedNodes).some(isValidInputNode) ||
+              Array.from(mutation.removedNodes).some(isValidInputNode),
+          );
+
+          if (hasMutationWithValidInputNode) {
+            updateDirtyCheck();
+          }
+        });
+
+        this.observer.observe(this.element, {
+          childList: true,
+          attributes: false,
+          subtree: true,
+        });
+      }, 1000 * 10);
+    }
+  }
+
+  disconnect(): void {
+    this.observer.disconnect();
+  }
+
+  beforeLeaving(event: BeforeUnloadEvent): void {
+    clearTimeout(this.isDirtyTimeout);
+    this.updateIsDirty();
+    const displayConfirmation =
+      !this.formSubmitted && (this.isDirty || this.isCommentsDirty);
+
+    if (displayConfirmation) {
+      event.preventDefault();
+      // eslint-disable-next-line no-param-reassign
+      event.returnValue = this.confirmationMessageValue;
+    }
+  }
+
+  getFormData() {
```

Comment:
It might make sense to get the other bit of code that resolves the form data to also call this method. 

Additionally, if you have a method that's starting with `get` it may make more sense to use class get methods. 

```suggestion
  get formData() {
```

This way you can call it simply by `this.formData`.